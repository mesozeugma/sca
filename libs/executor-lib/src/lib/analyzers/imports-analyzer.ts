import { AnalysisOutput } from '../analysis/analysis-output';
import { OpensearchIndex } from '../opensearch/opensearch-index';
import { readRepo } from '../read-repo';

import { Analyzer } from './analyzer';

export class ImportsAnalyzer implements Analyzer {
  async readRepo(repositoryPath: string) {
    try {
      const repositoryItems = await readRepo(repositoryPath);
      const repositoryImportItems = repositoryItems.filter(
        (value) => value.type === 'import'
      );
      return repositoryItems
        .map((e) =>
          AnalysisOutput.create(OpensearchIndex.ANALYSIS_RESULT_REPOSITORIES, e)
        )
        .concat(
          repositoryImportItems.map((e) =>
            AnalysisOutput.create(
              OpensearchIndex.ANALYSIS_RESULT_REPOSITORY_IMPORTS,
              e
            )
          )
        );
    } catch (e) {
      if (e instanceof Error) {
        return [
          AnalysisOutput.create(
            OpensearchIndex.ANALYSIS_RESULT_REPOSITORY_IMPORT_ERRORS,
            {
              errorMessage: e.message,
            }
          ),
        ];
      } else {
        throw e;
      }
    }
  }
}
