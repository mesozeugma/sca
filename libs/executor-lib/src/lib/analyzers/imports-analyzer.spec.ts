import { ImportsAnalyzer } from './imports-analyzer';

test('ImportsAnalyzer', async () => {
  const analyzer = new ImportsAnalyzer();
  const result = await analyzer.readRepo(
    './libs/executor-lib/code-samples/java/'
  );
  expect(result).toMatchSnapshot();
});
