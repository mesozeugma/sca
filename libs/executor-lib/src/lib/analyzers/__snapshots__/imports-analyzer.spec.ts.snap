// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImportsAnalyzer 1`] = `
[
  AnalysisOutput {
    "data": {
      "file": "HelloWorld.java",
      "imported": "java.util",
      "packageName": "code_samples",
      "type": "import",
    },
    "index": "public_sca_analysis_result_repositories",
  },
  AnalysisOutput {
    "data": {
      "className": "HelloWorld",
      "file": "HelloWorld.java",
      "packageName": "code_samples",
      "type": "package_or_class",
    },
    "index": "public_sca_analysis_result_repositories",
  },
  AnalysisOutput {
    "data": {
      "file": "nested/IsEven.java",
      "imported": "java.util",
      "packageName": "code_samples.nested",
      "type": "import",
    },
    "index": "public_sca_analysis_result_repositories",
  },
  AnalysisOutput {
    "data": {
      "className": "IsEven",
      "file": "nested/IsEven.java",
      "packageName": "code_samples.nested",
      "type": "package_or_class",
    },
    "index": "public_sca_analysis_result_repositories",
  },
  AnalysisOutput {
    "data": {
      "packageName": "code_samples",
      "type": "package_or_class",
    },
    "index": "public_sca_analysis_result_repositories",
  },
  AnalysisOutput {
    "data": {
      "packageName": "code_samples.nested",
      "type": "package_or_class",
    },
    "index": "public_sca_analysis_result_repositories",
  },
  AnalysisOutput {
    "data": {
      "file": "HelloWorld.java",
      "imported": "java.util",
      "packageName": "code_samples",
      "type": "import",
    },
    "index": "public_sca_analysis_result_repository_imports",
  },
  AnalysisOutput {
    "data": {
      "file": "nested/IsEven.java",
      "imported": "java.util",
      "packageName": "code_samples.nested",
      "type": "import",
    },
    "index": "public_sca_analysis_result_repository_imports",
  },
]
`;
